### PR TITLE
util: add cmp and as_bool

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @coocood @qiuyesuifeng PTAL

### CODE REVIEW

File: src/util/codec/datum.rs [hunk 1]
```diff
@@ -27,16 +29,137 @@ const UINT_FLAG: u8 = 4;
...
     Bytes(Vec<u8>),
-    MinValue,
-    MaxValue,
+    Min,
+    Max,
+}
+
+fn cmp_f64(l: f64, r: f64) -> Result<Ordering> {
+    l.partial_cmp(&r)
+     .ok_or_else(|| Error::InvalidDataType(format!("{} and {} can't be compare", l, r)))
```
Comment (coocood) [inline src/util/codec/datum.rs]: be compared.

**Comment by coocood** [comment]
LGTM

**Comment by BusyJay** [comment]
@ngaut @qiuyesuifeng PTAL

File: src/util/codec/convert.rs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+// limitations under the License.
+
+
+use std;
+
+use super::Result;
+
+/// `bytes_to_int` converts a byte arrays to an i64 in best effort.
+/// TODO: handle overflow.
+pub fn bytes_to_int(bytes: &[u8]) -> Result<i64> {
```
Comment (qiuyesuifeng) [inline src/util/codec/convert.rs]: Any need to rewrite this function? Can use standard library to do this?

File: src/util/codec/convert.rs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            (b"0", 0),
+            (b" 23a", 23),
+            (b"\t 23a", 23),
+            (b"\r23a", 0),
+            (b"1", 1),
+            (b"2.1", 2),
+            (b"23e10", 23),
+            (b"ab", 0),
+            (b"4a", 4),
+            (b"", 0),
```
Comment (qiuyesuifeng) [inline src/util/codec/convert.rs]: Add some more tests, like `-124`, `+124` and so on.

File: src/util/codec/datum.rs [hunk 4]
```diff
@@ -27,16 +29,137 @@ const UINT_FLAG: u8 = 4;
...
+            Datum::F64(f) => self.cmp_f64(f),
+            Datum::Bytes(ref bs) => self.cmp_bytes(bs),
+        }
+    }
+
+    fn cmp_i64(&self, i: i64) -> Result<Ordering> {
+        match *self {
+            Datum::I64(ii) => Ok(ii.cmp(&i)),
+            Datum::U64(u) => {
+                if i < 0 || u > i64::MAX as u64 {
```
Comment (qiuyesuifeng) [inline src/util/codec/datum.rs]: If i < 0, should be `Ok(Ordering::Less)`?

File: src/util/codec/datum.rs [hunk 5]
```diff
@@ -27,16 +29,137 @@ const UINT_FLAG: u8 = 4;
...
+    }
+
+    fn cmp_f64(&self, f: f64) -> Result<Ordering> {
+        match *self {
+            Datum::Null | Datum::Min => Ok(Ordering::Less),
+            Datum::Max => Ok(Ordering::Greater),
+            Datum::I64(i) => cmp_f64(i as f64, f),
+            Datum::U64(u) => cmp_f64(u as f64, f),
+            Datum::F32(ff) => cmp_f64(ff as f64, f),
+            Datum::F64(ff) => cmp_f64(ff as f64, f),
```
Comment (qiuyesuifeng) [inline src/util/codec/datum.rs]: No need to use `as f64`?

File: src/util/codec/datum.rs [hunk 6]
```diff
@@ -27,16 +29,137 @@ const UINT_FLAG: u8 = 4;
...
+        }
+    }
+
+    // as_bool converts an interface to a bool.
+    pub fn as_bool(&self) -> Result<bool> {
+        let b = match *self {
+            Datum::I64(i) => i != 0,
+            Datum::U64(u) => u != 0,
+            Datum::F32(f) => f.round() != 0f32,
+            Datum::F64(f) => f.round() != 0f64,
```
Comment (qiuyesuifeng) [inline src/util/codec/datum.rs]: For float64 type, is it OK to only using round function?
@coocood

File: src/util/codec/convert.rs (hunk 2 above)
Comment (BusyJay) [inline src/util/codec/convert.rs]: To keep compatible with mysql.

File: src/util/codec/datum.rs (hunk 4 above)
Comment (BusyJay) [inline src/util/codec/datum.rs]: No, if i < 0, then U >= 0 > i, so return Greater.

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**